sigstore/oidc.py

```diff
@@ -57,6 +58,145 @@ class _OpenIDConfiguration(BaseModel):
+        # We currently special-case a handful of these, and fall back
+        # on signing the "sub" claim otherwise.
+        identity_claim = _KNOWN_OIDC_ISSUERS.get(self.issuer)
+        if identity_claim is not None:
+            if identity_claim not in self._unverified_claims:
+                raise IdentityError(
+                    f"Identity token missing the required {identity_claim!r} claim"
+                )
+
+            self._identity = str(self._unverified_claims.get(identity_claim))
+        else:
+            try:
+                self._identity = str(self._unverified_claims["sub"])
+            except KeyError:
+                raise IdentityError("Identity token missing the required 'sub' claim")
+
+        # This identity token might have been retrieved directly from
+        # an identity provider, or it might be a "federated" identity token
+        # retrieved from a federated IdP (e.g., Sigstore's own Dex instance).
+        # In the latter case, the claims will also include a `federated_claims`
+        # set, which in turn should include a `connector_id` that reflects
+        # the "real" token issuer. We retrieve this, despite technically
+        # being an implementation detail, because it has value to client
+        # users: a client might want to make sure that its user is identifying
+        # with a *particular* IdP, which means that they need to pierce the
+        # federation layer to check which IdP is actually being used.
+        self._federated_issuer: str | None = None
+        federated_claims = self._unverified_claims.get("federated_claims")
+        if federated_claims is not None:
+            if not isinstance(federated_claims, dict):
+                raise IdentityError(
+                    "unexpected claim type: federated_claims is not a dict"
+                )
+
+            federated_issuer = federated_claims.get("connector_id")
+            if federated_issuer is not None:
+                if not isinstance(federated_issuer, str):
+                    raise IdentityError(
+                        "unexpected claim type: federated_claims.connector_id is not a string"
+                    )
+
+                self._federated_issuer = federated_issuer
+
+    @property
+    def identity(self) -> str:
+        """
+        Returns this `IdentityToken`'s underlying "subject".
+
+        Note that this is **not** always the `sub` claim in the corresponding
+        identity token: depending onm the token's issuer, it may be a *different*
+        claim, such as `email`. This corresponds to the Sigstore ecosystem's
+        behavior, e.g. in each issued certificate's SAN.
+        """
+        return self._identity
+
+    @property
+    def issuer(self) -> str:
+        """
+        Returns a URL identifying this `IdentityToken`'s issuer.
+        """
+        return self._issuer
+
+    @property
+    def expected_certificate_subject(self) -> str:
```

Comment:
This was `effective_certificate_subject` in a comment. I now you mentioned name is WIP, just pointing it out.

Reply:
> This was `effective_certificate_subject` in a comment. I know you mentioned name is WIP, just pointing it out.

Yeah, this was a shave on my part: "effective" implies something that's already occurred (to me), whereas "expected" implies some behavior that we expect from the Sigstore services (i.e., this is the value we *expect* them to use for the issuer extension).

Happy to rename if you prefer the other naming, however!